operator/controller/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,225 @@
+		&handler.EnqueueRequestForObject{}); err != nil {
+		return nil, nil, err
+	}
+
+	// Watch for the CoreDNS pod EndpointSlices to keep corednsEndpointsSliceCache synced.
+	// No reconcile requests should be generated.
+	if err := c.Watch(source.Kind(corednsEndpointsSliceCache, &discoveryv1.EndpointSlice{}),
+		handler.EnqueueRequestsFromMapFunc(func(context.Context, client.Object) []reconcile.Request {
+			return nil
+		})); err != nil {
+		return nil, nil, err
+	}
+
+	return c, []cache.Cache{dnsNameResolverCache, corednsEndpointsSliceCache}, nil
+}
+
+// Reconcile expects request to refer to an DNSNameResolver resource, and will do all the work to
+// keep the status of the resource updated.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// Ensure that the resolver is started before doing anything else.
+	r.startResolver.Do(func() {
+		r.resolver.Start()
+	})
+
+	controllerLog.Info("reconciling", "request", request)
+
+	// Get the DNSNameResolver resource.
+	dnsNameResolverObj := &ocpnetworkv1alpha1.DNSNameResolver{}
+	if err := r.dnsNameResolverCache.Get(ctx, request.NamespacedName, dnsNameResolverObj); err != nil {
+
+		// Check if the DNSNameResolver resource is deleted. If so, delete DNS names matching the DNSNameResolver resource
+		// from the resolver.
+		if errors.IsNotFound(err) {
+			r.resolver.Delete(request.Name)
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Get a copy of the DNSNameResolver resource.
+	dnsNameResolver := dnsNameResolverObj.DeepCopy()
+
+	// Check if the grace period is over for some of the IP addresses after the expiration of their respective TTLs. If so,
+	// remove those IP addresses and update the status of the resource.
+	if removalOfIPsRequired(&dnsNameResolver.Status) {
+		if err := r.client.Status().Update(ctx, dnsNameResolver); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Check if the TTLs of some of the IP addresses have expired. If so, requeue
+	// the reconcile request after the minimum remaining time until the grace
+	// period gets over among that of the IP addresses with expired TTLs.
+	if ttlExpired, remainingTime := reconcileRequired(&dnsNameResolver.Status); ttlExpired {
+		return reconcile.Result{Requeue: true, RequeueAfter: remainingTime}, nil
+	}
+
+	dnsName := string(dnsNameResolver.Spec.Name)
+	matchesRegular := !isWildcard(dnsName)
+
+	// Call the Add function of the resolver for the DNS name in the spec of the
+	// DNSNameResolver resource with empty resolved addresses. If it is a create
+	// event then this will ensure that the resolver sends the resolution request
+	// for the DNS name, so that the status of the object gets updated with the
+	// corresponding IP addresses.
+	r.resolver.Add(dnsName, []ocpnetworkv1alpha1.DNSNameResolverResolvedAddress{}, matchesRegular, request.Name)
```

Comment:
This `Add` call is required during the create event of the `DNSNameResolver` object. At that time the resolver needs to send the DNS lookup request for the DNS name in the object so that the CoreDNS plugin can update the status if the `DNSNameResolver` object.